### ISSUE TITLE
Explicit expert-axis coarse tiling cannot form a persistent expert loop

### ISSUE THREAD

**Issue by AdnanHoque** [torch-spyre/torch-spyre#3928]
## Summary

An explicit `spyre_hint` over the expert dimension cannot currently realize a dense expert FFN as one counted device loop.

For:

- `x: [T,H]`, shared by every expert
- gate/up weights: `[E,H,F]`
- down weights: `[E,F,H]`
- routing weights: `[E,T,1]`

the intended physical execution is:

1. stage `x` once;
2. execute one expert body in an `E`-trip device loop;
3. advance only the three weight operands and routing operand;
4. accumulate each weighted result into one `[T,H]` value in LX;
5. write that value once after the loop.

## Campaign status

Resolved and accepted. Every supporting mechanism and the final carried accumulator are in `main` through #4042 (`3358f39e`). The emitted structure, device numerics, and same-stack device performance have passed acceptance.

## Original compiler gaps

The implementation campaign identified four gaps:

1. When tiling reduces `E` to extent 1, dependency extraction removes the corresponding symbol before the original expert-to-expert address step is preserved.
2. Coarse tiling inserts temporary read copies for expert weights even when the matmul could read the advancing HBM slice directly.
3. A terminal reduction is represented as an ordinary tiled reduction rather than one loop-carried accumulator.
4. Scratchpad lifetime analysis sees the loop body only once and therefore cannot represent values that remain live across every loop iteration.

## Focused reproductions

The snippets below use the existing named-dimension test helpers to associate `E`, `T`, `H`, and `F` with tensor axes.

### 1. Preserve the expert address step and keep `x` fixed

```python
E, T, H, F = 2, 64, 64, 64
x = torch.randn(T, H, dtype=torch.float16)
w_backing = torch.randn(H, E, F, dtype=torch.float16)
w = w_backing.permute(1, 0, 2)  # [E,H,F]

with spyre_hint(num_tiles_per_dim={"E": E}):
    y = torch.matmul(x.unsqueeze(0), w)  # [E,T,F]
```

For the counted `E` loop, the generated operands must satisfy:

```text
x address step = 0
w address step = one expert weight slab
```

Tiling `E` to extent 1 must not erase the second fact.

### 2. Remove a temporary weight copy only after proving direct access

Use the same matmul above and inspect the post-layout IR/generated source.

Expected:

- the matmul reads the original graph-input weight argument;
- that argument carries the nonzero per-expert address step;
- no temporary read-copy buffer feeds the matmul.

If the original weight layout is incompatible with the matmul, the copy must remain. Copy removal is therefore a proven optimization, not an unconditional rewrite.

### 3. Carry one reduction value through the loop

```python
E, T, H = 2, 64, 64
values = torch.randn(E, T, H, dtype=torch.float16)

with spyre_hint(
    num_tiles_per_dim={"E": E},
    work_div={"T": 32},
):
    result = values.sum(dim=0)  # [T,H]
```

Expected generated structure:

```text
fill [T,H] accumulator once
for e in range(E):
    accumulator += values[e]
drain accumulator to result once
```

The per-tile local reduction has extent 1, so it contributes one value; the loop-carried add performs the real cross-expert reduction.

### 4. Preserve scratchpad lifetime across repeated loop trips

Allocator-level reproduction:

```python
persistent = LifetimeBoundBuffer(
    "persistent",
    size=64,
    uses=[0, 1],
    lifetime_end_override=5,
)
later = LifetimeBoundBuffer("later", size=64, uses=[2, 3])
```

Although the textual use lists do not overlap, `persistent` is live through the counted loop. The allocator must therefore assign the two buffers different addresses. The Python and native layout solvers must agree.

### 5. Require stable row ownership

Compile the reduction reproduction twice:

```python
# No ownership contract: retain the ordinary reduction path.
spyre_hint(num_tiles_per_dim={"E": E})

# Stable row ownership: loop-carried accumulation is legal.
spyre_hint(num_tiles_per_dim={"E": E}, work_div={"T": 32})
```

Only the second form may use one persistent accumulator. Every loop trip must assign a given token row to the same core.

## Expected result

For a hinted dense expert FFN, generated code should contain:

- one counted expert loop;
- one copy of the expert body;
- one invariant `x` staging operation;
- advancing gate, up, down, and routing operands;
- one non-aliasing `[T,H]` LX accumulator;
- no expert-expanded activation in HBM;
- one output drain after the loop.

Graphs without the explicit hint must retain their existing behavior.

## Landed implementation

- Capture input address advances before range division and use them only when a hinted dimension is later squeezed out.
- Preserve read copies through layout propagation, then remove a copy only after proving direct access is layout-compatible.
- Represent a terminal flat expert sum as fill, per-iteration add, and one post-loop drain.
- Extend the general scratchpad lifetime representation through counted loops.
- Require explicit stable row ownership before enabling the carried accumulator.

## Reduced complete FFN reproduction

This final case proves that the focused behaviors compose.

```python
E, T, H, F = 2, 64, 64, 64

# x: [T,H]
# gate_w/up_w: [E,H,F]
# down_w: [E,F,H]
# route: [E,T,1]

with spyre_hint(
    num_tiles_per_dim={"E": E},
    work_div={"T": 32},
):
    gate = torch.matmul(x.unsqueeze(0), gate_w)
    up = torch.matmul(x.unsqueeze(0), up_w)
    hidden = torch.nn.functional.gelu(gate, approximate="tanh") * up
    down = torch.matmul(hidden, down_w)
    result = (down * route).sum(dim=0)  # [T,H]
```

Acceptance requires device correctness against the ordinary PyTorch expression and the complete generated-program structure listed above.

The corresponding end-to-end regression in #3892 is `TestCoarseTileMoEBroadcastMatmulE2E.test_hinted_dense_expert_loop_keeps_one_body`.

## Accepted performance

These measurements cover two different boundaries:

- Expert FFN only, chunked versus persistent: `373.1 ms → 42.6 ms` (`8.8×`).
- Full Gemma 4 26B prefill, batch size 1 and sequence length 512: `6.4 s → 1.6 s` (`4.0×`).

The first isolates the expert FFN; the second is the end-to-end model prefill result.

## Tracking

Part of #3565.

All focused compiler issues and PRs are complete:

- #4035 → #4039: counted-loop scratchpad lifetime (merged);
- #4036 → #4040: operand address steps (merged);
- #4096: order-preserving work-division name translation (merged);
- #4075: allocation-only empty classification (merged);
- #4078: invariant activation staging (merged);
- #4037 → #4041: proven direct weight reads (merged);
- #4038 → #4042: carried terminal sum (merged).

Final compiler state:

```text
main@3358f39e contains the complete accepted dense expert-loop path
```

Structural, numerical-correctness, and same-stack device-performance acceptance passed. Prerequisite #3890 has merged into `main`.

**Comment by AdnanHoque** [comment]
The combined implementation has been split into four focused issues and draft PRs:

- [ ] #4035 — counted-loop scratchpad lifetime — #4039
- [ ] #4036 — preserve the operand step after size-one tiling — #4040
- [ ] #4037 — remove a temporary weight copy only after proving direct access — #4041
- [ ] #4038 — carry a terminal expert sum through the loop with stable row ownership — #4042

Dependency order: #4039 → #4040 → #4041 → #4042.

The final stacked tree is byte-for-byte identical to #3892. Each child issue points back to the focused reproduction in this issue; each replacement PR closes its corresponding child issue. #3892 is superseded by this sequence.

The red checks currently shown on #3892 are not compiler test failures. Both wheel jobs stopped in `astral-sh/setup-uv` while downloading uv 0.12.6 with HTTP 429, before the wheel or test suites ran.

**Comment by AdnanHoque** [comment]
**Updated implementation graph after review**

The earlier linear dependency note is superseded. The work now has three independent branches with one final integration point:

```text
main → #4039 counted-loop lifetime
main → #4040 address step → #4078 invariant staging → #4041 copy elision
main → #4075 empty classification → #4042 carried sum

integration = #4039 + #4041 leaf + #4042 leaf
```

**Ownership of each decision:**

1. #4039 keeps storage alive across repeated device-loop trips.
2. #4040 preserves a source address step before size-one tiling removes its symbol.
3. #4078 stages only proven-invariant inputs once before the loop.
4. #4041 removes an advancing HBM copy only after the final physical plan proves direct access equivalent.
5. #4075 classifies allocation-only `spyre.empty` as non-clobbering; real extern kernels remain barriers.
6. #4042 creates and finally verifies the loop-carried terminal sum and stable row ownership.

The branches deliberately do not share hidden policy. Each PR body records its direct prerequisite, what it owns, what it excludes, and the final integration contract. General dimension-lineage recovery for arbitrary positional rewrites is separate work; #4042 remaps names only at its known order-preserving reduction-collapse site.

**Comment by AdnanHoque** [comment]
**Final design-to-implementation audit**

The implementation now matches the approved split exactly:

```text
main → #4039 counted-loop lifetimes
main → #4040 address step → #4078 invariant staging → #4041 copy elision
main → #4075 empty classification → #4042 carried sum

integration = #4039 + #4041 leaf + #4042 leaf
```

**Owned behavior and dependencies:**

1. **#4039** extends storage lifetime across any counted loop and preserves that lifetime through LX-relayout handoffs. It owns no operand movement or carried-sum policy.
2. **#4040** preserves an advancing operand’s pre-division address step through size-one tiling. It supplies authoritative input to the selected immutable read-copy plan; it does not hoist or remove copies.
3. **#4078** hoists and compacts only loop-invariant staging copies. It depends on #4040 so advancing reads remain advancing.
4. **#4041** runs a named post-allocation transactional elision pass. It removes a read copy only after proving source identity, layout, address bounds, work division, output, and exact core-to-slice mapping unchanged. It therefore depends directly on #4040 and #4078.
5. **#4075** classifies allocation-only `SpyreEmptyFallback` as non-clobbering while preserving real extern-kernel barriers. It only lets a legal accumulator qualify for LX; it does not place or carry the accumulator.
6. **#4042** resolves the named row request before rewriting, gives fill/combine/drain one shared `CarriedReductionRecord`, and verifies the final scheduled ownership after LX demotion, fusion, and HBM-pool planning. Scheduler alpha-renaming is translated only for its known order-preserving pointwise stages; rank changes fail closed. It depends directly on #4075.

**Repository state:** all six heads contain current `main` (`85df435f`), are independently mergeable, DCO-signed, GPG-verified, and have zero unresolved review threads. #4039, #4040, #4078, #4041, and #4042 are fully green (138 successful checks, 3 intentional skips each). #4075’s implementation, focused tests, linters, and completed compiler suites pass; one unrelated broad device job was cancelled by a hardware RAS response timeout, and its required aggregate needs a maintainer rerun.

**Composition:** the three leaves compose without conflicts on a local, unpublished integration head. Ruff, `py_compile`, and `git diff --check` pass across the combined Python diff. No integration branch was published.

**Still required before claiming the feature complete:** run the combined E128 proof from one composed head: one counted loop, one stable LX accumulator, stable row ownership, no expert-expanded HBM intermediate, one drain, device/CPU numerical agreement, and matched performance. The current PRs provide the mechanisms and their focused proofs; they do not individually claim that final combined result.

**Comment by AdnanHoque** [comment]
**Updated implementation graph after production validation**

**Agreed correction:** The previous design treated general dimension lineage as deferred and kept a private name remap only at #4042's reduction-collapse site. The combined `E=128, T=512, F=704, work_div={"T": 32}` run showed that this was too narrow: the main matmul path can also remove the unit-sized `E` dimension and move the `T` request onto `F`.

**Implemented approach:**

1. #4096 captures logical output/reduction dimensions before coarse-tiling division.
2. It returns a checked, order-preserving old-symbol → new-symbol mapping; the caller moves `work_div_loop_info` through that mapping.
3. Generated read copies no longer inherit positional dimension names from differently shaped consumers.
4. Unsupported reorderings fail visibly instead of silently moving the request.
5. #4042 is restacked on #4096 and now uses this shared mechanism instead of its private positional remap.

```text
main → #4039 counted-loop lifetime
main → #4040 address step → #4078 invariant staging → #4041 copy elision
main (includes #4075) → #4096 dimension lineage → #4042 carried sum

integration = all three leaves
```

**Validation:** #4096 passes the full coarse-tiling suite (334 passed, 1 skipped). Refreshed #4042 focused validation passes 13 tests. The earlier combined v6 stack passed 532 focused compiler tests after a fresh native rebuild.

The design conclusion is now precise: order-preserving name translation through size-one coarse tiling is an integration prerequisite; arbitrary reorder-aware dimension recovery remains deferred.

**Comment by AdnanHoque** [comment]
Dependency refresh after the foundation merges:

```text
main includes #4075, #4039, and #4040
main → #4078 invariant staging → #4041 copy elision
main → #4096 dimension lineage → #4042 carried sum

integration = both open leaves on current main
```

This supersedes the older dependency graphs in the comment history. The four open PR descriptions and the active child issue bodies now carry the same graph.

**Comment by AdnanHoque** [comment]
**Full-stack refresh — 2026-08-28**

The implementation graph is now current and conflict-free:

```text
main@813a2980
  ├─ #4078@d4747d18 invariant staging
  │    └─ #4041@8f27e70f proven read-copy elision
  └─ #4042@13d37840 carried terminal sum
```

Merged prerequisites in `main`: #4075, #4039, #4040, and #4096.

**Refresh results:**

1. #4078 already contained current `main` and remains the active review leaf.
2. #4041 was rebuilt as one incremental commit on the current #4078 head. Its unit suite reports 350 passed/1 skipped, and all 8 focused device/compiler tests pass.
3. #4042 was rebuilt directly on current `main`. The refresh exposed and fixed the current work-division API adaptation: the carried row split is represented as a singleton legal split domain. Its unit suites report 376 passed/1 skipped/2 expected failures; all 9 focused carried-sum device/compiler tests pass.
4. Focused mypy, Ruff, formatting, and diff checks pass on both refreshed leaves. All refreshed commits are DCO-signed and GPG-signed.

Review order remains unchanged: #4078 first, then #4041; #4042 stays queued until the address branch is reviewed. Final integration still combines the #4041 and #4042 leaves and must re-run the complete E128 structural, numerical, and performance proof.

**Comment by AdnanHoque** [comment]
Compiler-stack update after #4078 merged:

```text
main@9aa76993 → #4041@6c649d46  (active review leaf)
main@9aa76993 → #4042@c0135099  (independent, queued leaf)
```

Merged mechanisms: #4075, #4039, #4040, #4096, and #4078. The PR descriptions, focused issues #4037/#4038, this tracker, and epic #3565 now carry the same graph. Final integration requires both remaining leaves plus structural, correctness, and same-stack device performance validation.

**Comment by AdnanHoque** [comment]
Compiler-stack update after #4041 merged:

```text
main@3855d111 includes #4075, #4039, #4040, #4096, #4078, and #4041
main → #4042@97fc388c  (active and final implementation leaf)
```

All address movement, staging, lifetime, and direct-read mechanisms are now landed. #4042 supplies the remaining loop-carried accumulator and final physical-ownership proof. After it lands, the campaign moves to combined structural, correctness, and same-stack device-performance acceptance.

**Comment by AdnanHoque** [comment]
Implementation and acceptance are complete in main through #4042 (`3358f39e`). All focused compiler issues are closed; structural, numerical-correctness, and same-stack device-performance gates passed.
